pynicotine/transfers.py

```diff
@@ -2117,43 +2113,46 @@ def get_upload_candidate(self):
 
                 uploading_users.add(user)
 
-                if user in queued_transfers:
-                    del queued_transfers[user]
+                if user in first_queued_transfers:
+                    del first_queued_transfers[user]
 
         oldest_time = None
         target_user = None
 
         if not round_robin_queue:
-            for user in queued_transfers:
-                if not privileged_queue or (privileged_queue and queued_users[user]):
-                    target_user = user
-                    break
+            # skip the looping below (except the cleanup) and get the first
+            # user of the highest priority we saw above
+            for user in first_queued_transfers:
+                if privileged_queue and not queued_users[user]:
+                    continue
+
+                target_user = user
+                break
 
         for user, update_time in list(self.user_update_times.items()):
+            # some cleanup, should probably log a warning as this case is
+            # probably a logic bug
             if user not in queued_users:
                 del self.user_update_times[user]
```

Comment:
We should also update the user time in the abort_transfer function (right before the call to check_upload_queue), otherwise you can end up with stuck queued uploads if you abort active uploads in a certain way. Aborting an upload should pass the torch to the next user in line, not start another upload for the same user.

Reply:
With your changes, I believe a user won't be added to queued_users if they have queued uploads and the user's first transfer in the list is an active one.

Reply:
> We should also update the user time in the abort_transfer function

If you say so, pushed.

> a user won't be added to queued_users if they have queued uploads and the user's first transfer in the list is an active one

So they will get cleared from user_update_times even though they still have uploads queued? I see what you mean, it should still work out if they get added back in when their current transfer ends but we maybe aren't updating it everywhere consistently.
I've changed it back. It changed the tests to show that if there are privileged users queued then non-privileged users won't get any slots even if the privileged users couldn't actually use them because they already have one active. I guess that is expected, if less than ideal.